File: exercises/python/8_capture_groups/evaluate.py
from unittest import TestCase
import re


from capture_groups import word_bookend_regex
from capture_groups import vi_swap_regex
from capture_groups import extract_log_parts_regex


def should_have_matched(regex, input_string, expected_match):
    return f'{regex} should have matched "{expected_match}"" for input string "{input_string}"'


def should_not_have_matched(regex, input_string):
    return f'{regex} matched input string "{input_string}" when it should not have.'


def empty_string_should_not_have_matched(regex):
    return f'{regex} should not have matched an empty input string'


class EvaluateWordBookendRegex(TestCase):
    regex_name = 'word_bookend_regex'

    def test_time_string(self):
        input_string = 'time time time'
        matches = re.search(word_bookend_regex, input_string)
        self.assertIsNotNone(
            matches,
            f'{self.regex_name} should have matched {input_string} but did not.')

    def test_I_string(self):
        input_string = 'I said no one is as lovely as I'
        matches = re.search(word_bookend_regex, input_string)
        self.assertIsNotNone(
            matches,
            f'{self.regex_name} should have matched {input_string} but did not.')

    def test_palindrome_string(self):
        input_string = 'madam I\'m Adam'
        matches = re.search(word_bookend_regex, input_string)
        self.assertIsNone(
            matches,
            f'{self.regex_name} matched {input_string} but it should not have.')

    def test_characters_not_words(self):
        input_string = 'ow, I don\'t know'
        matches = re.search(word_bookend_regex, input_string)
        self.assertIsNone(
            matches,
            f'{self.regex_name} matched {input_string} but it should not have.')

    def test_empty_string(self):
        input_string = ''
        matches = re.search(word_bookend_regex, input_string)
        self.assertIsNone(
            matches,
            f'{self.regex_name} matched {input_string} but it should not have.')


dir_list = """
.favorite_regexes.txt.swp
.practice.py.swp
.DS_Store
favorite_regexes.txt
practice.py
zippy.py
"""


class EvaluateViSwap(TestCase):
    regex_name = 'vi_swap_regex'

    def test_file_list(self):
        matches = re.findall(vi_swap_regex, dir_list)
        expected_match = ['favorite_regexes.txt', 'practice.py']
        self.assertEqual(
            matches,
            expected_match,
            should_have_matched(self.regex_name, dir_list, expected_match)
        )

    def test_empty_string(self):
        matches = re.findall(vi_swap_regex, '')
        self.assertEqual(
            matches,
            [],
            empty_string_should_not_have_matched(self.regex_name))

    def test_no_swpfiles(self):
        input_string = """file1.txt
file2.txt"""
        matches = re.findall(vi_swap_regex, input_string)
        self.assertEqual(
            matches,
            [],
            should_not_have_matched(self.regex_name, input_string))


class EvaluateLogPartsExtract(TestCase):

    def test_example_data(self):
        sample_data = """03Sep2020 04:55:38 This is a message
30Oct2020 23:44:01 This is another message"""
        match_objects = re.finditer(extract_log_parts_regex, sample_data)

        groupdicts = []
        if match_objects:
            for match in match_objects:
                groupdicts.append(match.groupdict())

        self.assertEqual(
            groupdicts,
            [
                {
                    'day': '03',
                    'month': 'Sep',
                    'year': '2020',
                    'time': '04:55:38',
                    'message': 'This is a message',
                },
                {
                    'day': '30',
                    'month': 'Oct',
                    'year': '2020',
                    'time': '23:44:01',
                    'message': 'This is another message',
                },
            ]
        )

File: exercises/javascript/8_capture_groups/evaluate.js
import { wordBookendRegex, viSwapRegex, extractLogPartsRegex } from '.';

describe('EvaluateWordBookendRegex', function () {
  it('matches time_string', function () {
    const match = wordBookendRegex.test('time time time');
    expect(match).toBe(true);
  });
  it('matches I_string', function () {
    const match = wordBookendRegex.test('I said no one is as lovely as I');
    expect(match).toBe(true);
  });
  it('does not match palindrome_string', function () {
    const match = wordBookendRegex.test("madam I'm Adam");
    expect(match).toBe(false);
  });
  it('does not match characters_not_words', function () {
    const match = wordBookendRegex.test("ow, I don't know");
    expect(match).toBe(false);
  });
  it('does not match empty_string', function () {
    const match = wordBookendRegex.test('');
    expect(match).toBe(false);
  });
});

const dirList =
  '.favorite_regexes.txt.swp\n.practice.py.swp\n.DS_Store\nfavorite_regexes.txt\npractice.py\nzippy.py';

describe('EvaluateViSwap', function () {
  it('finds swap files in file_list', function () {
    // the only way to get groups from multiple matches is to use RegExp.prototype.exec
    // (string.prototype.match will only return full matches, not groups)
    let match;
    const matches = [];
    while ((match = viSwapRegex.exec(dirList))) matches.push(match[1]); // match[0] is the complete match; match[1] is the group
    expect(matches).toEqual(['favorite_regexes.txt', 'practice.py']);
  });
  it('does not match empty_string', function () {
    const match = viSwapRegex.exec('');
    expect(match).toBe(null);
  });
  it('does not match no_swpfiles', function () {
    const match = viSwapRegex.exec('file1.txt\nfile2.txt');
    expect(match).toBe(null);
  });
});

describe('EvaluateExtractLogPartsRegex', function () {
  it('extracts the proper parts from the example', function () {
    const exampleLog =
      '03Sep2020 04:55:38 This is a message\n30Oct2020 23:44:01 This is another message';
    const matches = [...exampleLog.matchAll(extractLogPartsRegex)];
    const namedMatches = matches.map((match) => match.groups);
    expect(namedMatches).toEqual([
      {
        day: '03',
        month: 'Sep',
        year: '2020',
        time: '04:55:38',
        message: 'This is a message',
      },
      {
        day: '30',
        month: 'Oct',
        year: '2020',
        time: '23:44:01',
        message: 'This is another message',
      },
    ]);
  });
});
